### ISSUE TITLE
Prometheus data source in Grafana returns incorrect and truncated data

### ISSUE THREAD

**Issue by readmodifywrite** [VictoriaMetrics/VictoriaMetrics#4832]
### Describe the bug

VM 1.93.0

I have inserted the following test data via the JSON line protocol:

```
data = {
        "metric": {
            '__name__': 'meow',
        },
        'values': [1, 2, 3, 4, 3, 2, 1],
        'timestamps': [
                1692048000000, 
                1692048001000, 
                1692048002000, 
                1692048003000,
                1692048004000,
                1692048005000,
                1692048006000,
        ]
    }
```

The timestamps start at 2023-08-14T21:20:00.000Z and increment by 1 second each.  This data set covers a total of 7 seconds.

When I query in Grafana via the Prometheus data source, I get this:

![2023-08-14_16-44](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/3618507/7c690242-8247-480d-b829-d39636cb7780)

Only the first and latest data points are returned.  Additionally, the timestamp on the last datapoint is incorrect.  It should be at 6 seconds, not 15.  There should be 5 other datapoints in between, and they are missing.

If I do a raw JSON dump, it's all actually in there, it just doesn't show up via the Prometheus data source.

I also tried to do this with an Influx data source, which doesn't even report that my metric exists at all (another issue has been filed for this).

I'm used to InfluxDB where all of the data comes back on the correct timestamps, and my understanding was that VM is supposed to be able to function as a replacement for Influx (which is still my hope).  However, I need the queries in Grafana to come back with all of the data and the correct timestamps, just like Influx, and so far I am not observing that behavior.

### To Reproduce

Insert data.  Query in Grafana.

### Version

1.93.0

### Logs

I didn't see any logging output associated with this query.

**Comment by readmodifywrite** [comment]
It looks like I can set the "step" field in Grafana to 1 second, and the low level granularity is restored.  I do not understand why throwing away a ton of data points and then moving the final data point forward in time by 9 seconds is a sensible default.

Additionally, setting the step field still returns  corrupted data:

![2023-08-14_16-57](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/3618507/331f7121-ed8b-4625-a16b-0b94e1e7cc1b)

This data set ends at 16:20:06.  The final data point is repeated on 1 second intervals until 16:20:10.   This is 4 data points inserted that are not present in the actual data set.  That is not acceptable for my use case - I need to see the actual data, including subsecond resolution with no missing points, correct timestamps, and no data inserted to pad out the timing to a 10 second offset.

Influx doesn't have these kinds of issues, as I've already noted.  Perhaps I'm doing something wrong here, but is there a way to get VM to actually provide the same kind of time series functionality that Influx does?  The docs really seem to push this claim, but I'm just not seeing it play out.

**Comment by hagen1778** [comment]
Hello @readmodifywrite! I'd recommend reading https://docs.victoriametrics.com/guides/migrate-from-influx.html#query-data.

tl;dr; VictoriaMetrics or Prometheus querying API are supposed to show system **state**, not system **events**. That's why param `step` exists. It basically tells to API "show me how system state was looking like each, let's say, 1m during last 1h". It is quite different to a traditional "show me everything that happened to the system for last 1h".

> This data set ends at 16:20:06. The final data point is repeated on 1 second intervals until 16:20:10. This is 4 data points inserted that are not present in the actual data set.

The same reason as above. API plots ephemeral (non existing) data points continuing to show unchanged system state. See more about this https://docs.victoriametrics.com/keyConcepts.html#range-query.

I'd also recommend watching the talk of mine explaining more about time series analysis specifics here https://www.youtube.com/watch?v=_zORxrgLtec

**Comment by readmodifywrite** [comment]
OK, so it sounds like VM is really a Prometheus style DB and doesn't really do what Influx does.  The docs really make it sound like it can do both jobs (which would have been awesome for my use case).

I think the Prometheus style of query makes a lot of sense for server monitoring, but in my use case it doesn't really work well with IoT sensor data.  I need to see all of the data points plotted without any editorializing from the database.  Old versions of Influx can do that just fine.  What goes in is what I see.

I'm here at all because Influx has really been a mess over the last several years and the older versions (with waning support) are the most suitable for modest IoT systems that use Grafana as a front end.  I still would really like to replace it, but I'm not sure VM is going to cut it.

Is there any plan to add Influx style queries for the future?  I don't care how the data source works - Prom or Influx or whatever.  I just want Influx behavior since that's what I'm trying to replace.

I guess my tl;dr is this:
I actually want system *events* for my IoT data, which Influx gives me.
I want system *state* for server monitoring, which Prom gives me.

The docs really make it sound like VM is supposed to be able to do both - unless maybe there are a ton of people using Influx for the job Prom is better suited for.  But a lot of the use cases for Influx (like push based IoT sensor data) are a very poor match for Prom.


Well, anyway, thanks for responding.  I guess I'll keep my ancient Influx instances a live as long as I can.  If VM ever supports Influx style queries it would be a very compelling option for me!

And FWIW it is definitely easier to use for Prom style monitoring so I'll probably continue using it for that over Prom.

**Comment by hagen1778** [comment]
> Is there any plan to add Influx style queries for the future? I don't care how the data source works - Prom or Influx or whatever. I just want Influx behavior since that's what I'm trying to replace.

We don't have such plans yet. However, you can try applying the following hack from the docs:
> VictoriaMetrics fills the gaps on the graph assuming time series are always continuous and not discrete. To limit the interval on which VictoriaMetrics will try to fill the gaps, set -search.setLookbackToStep command-line flag. This limits the gap filling to a single step interval passed to [/api/v1/query_range](https://docs.victoriametrics.com/keyConcepts.html#range-query). This behavior is close to InfluxDB data model.

If it doesn't help and VM still is not good fit for your case, I'd recommend taking a look at https://github.com/ClickHouse/ClickHouse - OS free DBMS which performs way better than influx.

**Comment by hagen1778** [comment]
Btw, could you please point me to the lines in docs which made you think VM can do Influx query language? I'd prefer to make it more clear for future readers.

**Comment by readmodifywrite** [comment]
Thanks for the recommendation, I'll check out ClickHouse.

I think there are a couple sources of my confusion, and not all of them are VM specific, so there's only so much you can really do.  But I'll try to explain how I got here.

1. In general, the industry has been really overloading the concept of a time-series database to mean slightly different but substantially overlapping things.  There are really two different styles of database, what I'll call the Prometheus style and the Influx style.  Both are doing time series metrics, but the way they do it and the use cases they are suitable for are rather different.  This is starting to make it difficult to shop around for a time series database.  The Prometheus use case is really targeting server/cloud/service monitoring ala the original RRD and Graphite, but with a really convenient pull based ingestion method.  The Influx style is for pushed based metrics coming in from sensor data - often real world physical sensors (another way to think about it is telemetry, NASA style. We are getting remotely recorded sensor data and want to see the raw data.  If data is missing, then it's missing. Etc.).  The two styles don't just differ in the ingestion method, as the query engines are clearly also different to reflect the differing needs between the two uses cases.

Back when it was just Prometheus, it was easier to tell the difference, but now everyone seems to want to get in on the Prometheus style (which is fine, and I'll point out that I think VM is easier to set up and use than Prom for that use case) - but a lot of us still need an Influx style telemetry push based database.

Something I'd like to point out is that the combination of Grafana + Influx 1.8 is incredibly easy to use for doing IoT sensor analytics.  Influx really screwed this up on the 2.x branch and seems to be heading down quite a technical debt ridden rabbit hole for 3.x, with no apparently plan to go back to what really worked very well (and I'm talking about my use case here, so it's different strokes for different folks).  There is potentially a gap in the market here.  There certainly is for me, at least.

So to summarize this point: the Prometheus model is great for service monitoring but poorly suited to raw telemetry/sensor data - especially higher rate data (sub second).  The industry doesn't have clear terms of art to describe the difference and that's part of what makes it hard to shop for what I need.


2.  For VM specifically: VM really pushes support for a wide variety of ingestion protocols, and prominently claims InfluxDB as one of them.  (Side note: I actually originally tried using my Influx client code to push to VM, but it didn't work.  I ended up using the JSON line protocol instead.  I didn't bother with a ticket because it was actually simpler this way, and I did get it to work.).  So that's part of it for me.  Why support ingestion for Influx if you aren't going to support querying the database like you would with Influx?  I'm using Influx and Prom - both for the different jobs they are suited for.  It sounded like VM could do both in one package while also offering better performance.  How could I say no to that?

Given the above, I was assuming orthogonality.  If you accept Influx data (and originally I assumed the Influx client - the Python version in my case - would also work), I assumed that it would act as an Influx data *source* as well.  Basically, a drop-in replacement for Prometheus AND Influx all in one, with better performance, better tooling, and better API support.  That's a slam dunk!

So I think probably VM could clarify that it accepts ingestion from sources like Influx, but that it isn't designed to query the data in the same way and can't replicate all of the use cases that Influx does.  Data goes in Influx style, but only comes out Prometheus style.  And ideally the industry at large either figures out how to delineate between the two, or achieve some kind of harmony where both can credibly do the same job.  The difference is in the ingestion and query models, but the storage engines (and quite like the actual query engines) could probably be nearly the same.  They just unfortunately aren't at this time.

While you did mention that there aren't plans to support Influx style queries, it really could be awesome if this was an option in the future.  VM's tooling and storage model looks well thought out and actively supported, and I can't be the only engineer who's hungry for a compelling replacement for Influx.

In the meantime, VM still looks great to replace Prometheus and probably Loki as well.  So I'll probably keep moving forward with that, and investigate ClickHouse (or just keep Influx 1.8 alive for as long as I can!).


Since this isn't really a bug and isn't a feature you are actively considering, I'll close this ticket out.  Thanks for the quick responses!

**Comment by readmodifywrite** [comment]
FWIW, I did try the setLookbackToStep option.  It seems to help *if* you also set the min step field in the query from auto to 1 (not sure how well that will work with sub-second data though).  However, it's frankly not great that the default query is still "wrong" and requires an extra field on every query to get to what I need.

Is there a way to configure the min step field to just always be set to return the full data set?  I realize this may be an issue with Grafana's Prometheus data source (and the VM data source as well) and not the VM query engine - but again I'll point out that the Influx data source just correctly returns the data set by default with no modifications - which is what I need.  If there's a way to get VM/Grafana to do that by default without having to hand tweak every query, then maybe VM could still work out for me.

Part of my use case for my IoT queries is that casual users will be interacting with the system.  I want to give them the power and ease of Grafana, but I can't expect them to understand the inner workings of the query model.  I need the default time based range query to return the complete data set verbatim (which is what Influx does).  This is another area where the Prometheus style can get away with extra steps: Prom users are generally engineers who are monitoring their own infrastructure.  I want to help non-engineers analyze their sensor data with minimal fuss or learning curve.

**Comment by hagen1778** [comment]
> If there's a way to get VM/Grafana to do that by default without having to hand tweak every query, then maybe VM could still work out for me.

Unfortunately, I'm not aware of easy to do this. Grafana automatically sets `step` param for all sent queries and VM API respects that. Probably, you can have a proxy between Grafana and VM to replace `step` param with the desired value.
